vortex-serde/src/file/file_writer.rs

```diff
@@ -0,0 +1,291 @@
+            .copied()
+            .unwrap_or(0u64);
+        let mut byte_offsets = vec![self.msgs.tell()];
+        let mut row_offsets = vec![column_row_offset];
+
+        while let Some(chunk) = stream.try_next().await? {
+            row_offsets.push(
+                row_offsets
+                    .last()
+                    .map(|off| off + chunk.len() as u64)
+                    .expect("Row offsets should be initialized with a value"),
+            );
+            self.msgs.write_batch(chunk).await?;
+            byte_offsets.push(self.msgs.tell());
+        }
+
+        Ok(ChunkLayout {
+            byte_offsets,
+            row_offsets,
+        })
+    }
+
+    fn merge_chunk_offsets(&mut self, column_idx: usize, chunk_pos: ChunkLayout) {
+        if let Some(chunk) = self.column_chunks.get_mut(column_idx) {
+            chunk.byte_offsets.extend(chunk_pos.byte_offsets);
+            chunk.row_offsets.extend(chunk_pos.row_offsets);
+        } else {
+            self.column_chunks.push(chunk_pos);
+        }
+    }
+
+    async fn write_metadata_arrays(&mut self) -> VortexResult<StructLayout> {
+        let DType::Struct(..) = self.dtype.as_ref().expect("Should have written values") else {
+            unreachable!("Values are a structarray")
+        };
+
+        let mut column_layouts = Vec::with_capacity(self.column_chunks.len());
+
+        for mut chunk in mem::take(&mut self.column_chunks) {
+            let mut chunks = Vec::new();
+
+            let len = chunk.byte_offsets.len() - 1;
+            let byte_counts = chunk
+                .byte_offsets
+                .iter()
+                .skip(1)
+                .zip(chunk.byte_offsets.iter())
+                .map(|(a, b)| a - b)
+                .collect_vec();
+
+            chunks.extend(
+                chunk
+                    .byte_offsets
+                    .iter()
+                    .zip(chunk.byte_offsets.iter().skip(1))
+                    .map(|(begin, end)| Layout::Flat(FlatLayout::new(*begin, *end))),
+            );
+            let row_counts = chunk
+                .row_offsets
+                .iter()
+                .skip(1)
+                .zip(chunk.row_offsets.iter())
+                .map(|(a, b)| a - b)
+                .collect_vec();
+            chunk.byte_offsets.truncate(len);
+            chunk.row_offsets.truncate(len);
+
+            let metadata_array = StructArray::try_new(
+                [
+                    "byte_offset".into(),
+                    "byte_count".into(),
+                    "row_offset".into(),
+                    "row_count".into(),
+                ]
+                .into(),
+                vec![
+                    chunk.byte_offsets.into_array(),
+                    byte_counts.into_array(),
+                    chunk.row_offsets.into_array(),
+                    row_counts.into_array(),
+                ],
+                len,
+                Validity::NonNullable,
+            )?;
```

Comment:
This logic beyond extending chunks byte offsets is highly suspect. We can't reuse these tables in ChunkedArrayReader which is a bad thing. I wonder if we need to store multiple metadata tables since the offset tables are going to be 1 row longer than stat tables. OR the byte/row offsets live in layouts and these tables are pure metadata.